language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -104,222 +135,266 @@ kindOfTypeBuiltin TyByteString = sizeToType
     :: ann
-    -> Type TyNameWithKind ann  -- ^ A pattern functor.
-    -> Kind ()                  -- ^ @k@.
+    -> Type TyName ann  -- ^ A pattern functor.
+    -> Kind ()          -- ^ @k@.
     -> TypeCheckM ann ()
 kindCheckPatternFunctorM ann pat k =
     kindCheckM ann pat $ KindArrow () (KindArrow () k (Type ())) (KindArrow () k (Type ()))
 
--- | Check a 'Type' against a 'Kind'.
-kindCheckM :: a -> Type TyNameWithKind a -> Kind () -> TypeCheckM a ()
-kindCheckM x ty k = do
-    tyK <- kindOf ty
-    when (tyK /= k) $ throwError (KindMismatch x (void ty) k tyK)
-
 -- | Apply a 'TypeBuiltin' to a 'Size' and wrap in 'NormalizedType'.
 applySizedNormalized :: TypeBuiltin -> Size -> NormalizedType tyname ()
 applySizedNormalized tb = NormalizedType . TyApp () (TyBuiltin () tb) . TyInt ()
 
 dummyUnique :: Unique
 dummyUnique = Unique 0
 
-dummyTyName :: TyNameWithKind ()
-dummyTyName = TyNameWithKind (TyName (Name ((), Type ()) "*" dummyUnique))
+dummyTyName :: TyName ()
+dummyTyName = TyName (Name () "*" dummyUnique)
 
 dummyKind :: Kind ()
 dummyKind = Type ()
 
-dummyType :: Type TyNameWithKind ()
+dummyType :: Type TyName ()
 dummyType = TyVar () dummyTyName
 
 -- | Look up a 'DynamicBuiltinName' in the 'DynBuiltinNameTypes' environment.
-lookupDynamicBuiltinName :: a -> DynamicBuiltinName -> TypeCheckM a (NormalizedType TyNameWithKind ())
+lookupDynamicBuiltinName :: a -> DynamicBuiltinName -> TypeCheckM a (NormalizedType TyName ())
 lookupDynamicBuiltinName ann name = do
     dbnts <- asks $ unDynamicBuiltinNameTypes . _typeConfigDynBuiltinNameTypes
     case Map.lookup name dbnts of
-        Nothing    ->
+        Nothing ->
             throwError $ UnknownDynamicBuiltinName ann (UnknownDynamicBuiltinNameErrorE name)
-        Just quoTy -> do
-            ty <- liftQuote quoTy
-            annotateNormalizeType ann (DynBuiltinName () name) ty
+        Just ty -> rename ty
 
 -- | Get the 'Type' of a 'Constant' wrapped in 'NormalizedType'.
-typeOfConstant :: Constant a -> NormalizedType TyNameWithKind ()
+typeOfConstant :: Constant a -> NormalizedType TyName ()
 typeOfConstant (BuiltinInt  _ size _) = applySizedNormalized TyInteger    size
 typeOfConstant (BuiltinBS   _ size _) = applySizedNormalized TyByteString size
 typeOfConstant (BuiltinSize _ size)   = applySizedNormalized TySize       size
 typeOfConstant (BuiltinStr _ _)       = NormalizedType $ TyBuiltin () TyString
 
-typeOfBuiltin :: Builtin a -> TypeCheckM a (NormalizedType TyNameWithKind ())
-typeOfBuiltin (BuiltinName    ann name) = normalizedAnnotatedTypeOfBuiltinName ann name
+typeOfBuiltin :: Builtin a -> TypeCheckM a (NormalizedType TyName ())
+-- We have a weird corner case here: the type of a 'BuiltinName' can contain 'TypedBuiltinDyn', i.e.
+-- a static built-in name is allowed to depend on a dynamic built-in type which are not required
+-- to be normalized. For dynamic built-in names we store a map from them to their *normalized types*,
```

Comment:
Hm, why not require dynamic builtin types to be normalized?

Reply:
AFAICT if we require `getTypeEncoding` to return a `Quote NormalizedType` then it should all bubble up fine. All the type scheme encoding stuff just creates foralls and type arrows, no applications or anything.

Reply:
> All the type scheme encoding stuff just creates foralls and type arrows, no applications or anything.

I'll likely add applications in future, but yeah, `TypeScheme` are probably going to stay normalized. I'm just not sure whether `getTypeEncoding` returning `Quote NormalizedType` is a good thing, because we obviously don't want the instance writer to manually normalize everything, hence we need to provide the type normalization procedure, but type normalization requires kind checking and I'm not sure if this makes everything mutually recursive or recursive. I'm planning to try that out just to see how it goes, but certainly not in this PR.

Reply:
Hm yes, I can see that that's awkward.